src/runtime/GlobalObjectBuiltinString.cpp

```diff
@@ -907,21 +907,68 @@ static Value builtinStringToLocaleUpperCase(ExecutionState& state, Value thisVal
     return builtinStringToUpperCase(state, thisValue, argc, argv, isNewExpression);
 }
 
-static Value builtinStringTrim(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
+enum StringTrimWhere : unsigned {
+    START,
+    END,
+    STARTEND
+};
+
+static Value builtinStringTrimString(ExecutionState& state, Value thisValue, StringTrimWhere where)
 {
+    // Let str be ? RequireObjectCoercible(string).
+    // Let S be ? ToString(str).
     RESOLVE_THIS_BINDING_TO_STRING(str, String, trim);
+
     int64_t s, e;
-    for (s = 0; s < (int64_t)str->length(); s++) {
-        if (!EscargotLexer::isWhiteSpaceOrLineTerminator((*str)[s]))
-            break;
+    if (where == START)
+        e = str->length() - 1;
+    if (where == END)
+        s = 0;
+
+    // If where is "start", let T be the String value that is a copy of S with leading white space removed.
+    // Trim beginning if start, or start+end
+    if (where == START || where == STARTEND) {
+        for (s = 0; s < (int64_t)str->length(); s++) {
+            if (!EscargotLexer::isWhiteSpaceOrLineTerminator((*str)[s]))
+                break;
+        }
     }
-    for (e = ((int64_t)str->length()) - 1; e >= s; e--) {
-        if (!EscargotLexer::isWhiteSpaceOrLineTerminator((*str)[e]))
-            break;
+    // Else if where is "end", let T be the String value that is a copy of S with trailing white space removed.
+    // Trim ending if end or start+end
+    if (where == END || where == STARTEND) {
+        for (e = ((int64_t)str->length()) - 1; e >= s; e--) {
+            if (!EscargotLexer::isWhiteSpaceOrLineTerminator((*str)[e]))
+                break;
+        }
     }
+    // Return T.
     return new StringView(str, s, e + 1);
 }
 
+static Value builtinStringTrim(ExecutionState& state, Value thisValue, size_t argc, Value* argv, bool isNewExpression)
+{
+    // Let S be this value.
+    // Return ? TrimString(S, "start+end").
+    RESOLVE_THIS_BINDING_TO_OBJECT(S, String, trim);
```

Comment:
IMO this is unnecessary since `builtinStringTrimString` starts with `RESOLVE_THIS_BINDING_TO_STRING` (same as below).

Reply:
I can remove them if you want.I thought about it too, but just went with the ecma specs.